src/python/pants/backend/rust/util_rules/toolchains.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import dataclasses
+import os
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.rust.subsystems.rust import RustSubsystem
+from pants.core.util_rules.system_binaries import (
+    BinaryPath,
+    BinaryPathRequest,
+    BinaryPaths,
+    BinaryPathTest,
+)
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.internals.native_engine import Digest
+from pants.engine.internals.selectors import Get
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessCacheScope, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.util.logging import LogLevel
+
+
+class RustupBinary(BinaryPath):
+    """Path to the rustup binary used to select and gain access to Rust toolchains."""
+
+
+@dataclass(unsafe_hash=True)
```

Comment:
This is something I discussed with @tdyas.

We are using rustup which is a common way of installing rust to tell something about which toolchain is currently in use.

Reply:
Right, but this specific code, are you copying from somewhere else in the Pants codebase? Otherwise I'm confused by things like using the `unsafe_hash` etc...)

Reply:
Yes, `unsafe_hash` is in use in several other places to ensure `__hash__` is generated when the dataclass is not frozen. Here the dataclass is not frozen because we have a convenience constructor.

As per the [dataclasses documentation](https://docs.python.org/3/library/dataclasses.html), contrary to the what an ordinary interpretation of the word "unsafe" might be, `unsafe_hash` does *not* mean "define a hash which itself is unsafe," `unsafe_hash` instead just means to define the `__hash__` special method which is not defined automatically when the dataclass is not frozen. Thus, "unsafe" here means "Yes Python, I know what I am doing and all of the dataclass attributes are hashable, please define the __hash__ special m method."

Reply:
It can still be frozen and have a special constructor. Why aren't we doing that?

Reply:
```
E       pants.engine.internals.scheduler.ExecutionError: 1 Exception encountered:
E
E       Engine traceback:
E         in select
E           ..
E         in pants.backend.rust.lint.rustfmt.rules.rustfmt_fmt
E           Format with rustfmt
E
E       Traceback (most recent call last):
E         File "/private/var/folders/4_/b_1y6cg128d4rtqnqmxrm9z00000gn/T/pants-sandbox-XhZ6hj/src/python/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
E           res = rule.send(arg) if err is None else rule.throw(throw or err)
E         File "/private/var/folders/4_/b_1y6cg128d4rtqnqmxrm9z00000gn/T/pants-sandbox-XhZ6hj/src/python/pants/backend/rust/lint/rustfmt/rules.py", line 48, in rustfmt_fmt
E           RustToolchainProcess(
E         File "/private/var/folders/4_/b_1y6cg128d4rtqnqmxrm9z00000gn/T/pants-sandbox-XhZ6hj/src/python/pants/conftest.py", line 59, in new__init__
E           old__init__(self, *args, **kwargs)
E         File "/private/var/folders/4_/b_1y6cg128d4rtqnqmxrm9z00000gn/T/pants-sandbox-XhZ6hj/src/python/pants/backend/rust/util_rules/toolchains.py", line 54, in __init__
E           self.binary = binary
E         File "<string>", line 4, in __setattr__
E       dataclasses.FrozenInstanceError: cannot assign to field 'binary'

src/python/pants/engine/internals/scheduler.py:500: ExecutionError
```

Reply:
The above happens given a change to frozen

Reply:
See https://docs.python.org/3/library/dataclasses.html#frozen-instances
and the other frozen dataclasses. You have to use `object.__setattr__(self, "attrname", <value>)`

Reply:
Also what's the difference between this class and `Process`? Could you just inherit? Seems like there's a very large overlap.

Reply:
> See https://docs.python.org/3/library/dataclasses.html#frozen-instances
and the other frozen dataclasses. You have to use object.__setattr__(self, "attrname", <value>)

I recall maybe that this was a general refactor made in the code base to switch to `object.__setaddr__`?

Reply:
> Also what's the difference between this class and Process? Could you just inherit? Seems like there's a very large overlap.

`RustToolchainProcess` abstracts over the particular Rust toolchain so other Rust rules do not need to be concerned with what toolchain to use.

It is a similar pattern to [`GoSdkProcess`](https://github.com/pantsbuild/pants/blob/3ce90948d2dc49dd525ca4a10daeaba98068a8e0/src/python/pants/backend/go/util_rules/sdk.py#L24) and [`JvmProcess`](https://github.com/pantsbuild/pants/blob/3ce90948d2dc49dd525ca4a10daeaba98068a8e0/src/python/pants/jvm/jdk_rules.py#L316). All of them abstract over the underlying SDK or toolchain.

Reply:
Also inheritance would not be favored here since the conversion `RustToolchainProcess` -> `Process` is done by a rule and the engine requires they be distinct types.

Reply:
The engine is ambivalent to subprocess. Let's just file an issue to clean this us repo-wide.

Then this just tracks making it frozen.

Reply:
Clean it up, yeah. If that means removal, that's probably even better 👍

Reply:
You should open a discussion then, because those types exist so that most of the rules within each of their respective backends do not need to reason about the location of the respective language SDK.

Reply:
For this PR, `RustToolchainProcess` should remain.

Reply:
Is resolved now. Let me know if you have anything further

Reply:
Bueno. 👍